src/FontManager.cc

```diff
@@ -1,251 +1,265 @@
 template<bool async>
-NAN_METHOD(findFont) {
-  if (info.Length() < 1 || !info[0]->IsObject() || info[0]->IsFunction())
-    return Nan::ThrowTypeError("Expected a font descriptor");
+Napi::Value FontManagerInterface::findFont(const Napi::CallbackInfo& info) {
+  Napi::Env const &env = info.Env();
+  FontManagerImpl *impl = env.GetInstanceData<FontManagerImpl>();
 
-  Local<Object> desc = info[0].As<Object>();
-  FontDescriptor *descriptor = new FontDescriptor(desc);
+  if (info.Length() < 1 || !info[0].IsObject()) {
+    return throwError<async>(env, Napi::TypeError::New(env, "Expected a font descriptor"));
+  }
+  Napi::Object desc = info[0].As<Napi::Object>();
+  FontDescriptor *descriptor = new FontDescriptor(env, desc);
 
   if (async) {
-    if (info.Length() < 2 || !info[1]->IsFunction())
-      return Nan::ThrowTypeError("Expected a callback");
-
-    AsyncRequest *req = new AsyncRequest(info[1]);
-    req->desc = descriptor;
-    uv_queue_work(uv_default_loop(), &req->work, findFontAsync, (uv_after_work_cb) asyncCallback);
-
-    return;
+    auto worker = new PromiseAsyncWorker<FontDescriptor>(env, [=](FontDescriptor **result) {
+      long error = impl->findFont(result, descriptor);
+      // need to clean up descriptor somewhere
+      delete descriptor;
+      return error;
+    });
+    worker->Queue();
+
+    return worker->GetPromise();
   } else {
     FontDescriptor *result = NULL;
-    long error = findFont(&result, descriptor);
+    long error = impl->findFont(&result, descriptor);
     if (error != 0) {
-      return Nan::ThrowError(Nan::ErrnoException(error));
+      Napi::Error::New(env).ThrowAsJavaScriptException();
+      return env.Null();
     }
 
-    Local<Value> res = wrapResult(result);
+    Napi::Value res = wrapResult(env, result);
     delete descriptor;
-    info.GetReturnValue().Set(res);
+    return res;
   }
 }
 
-void substituteFontAsync(uv_work_t *work) {
-  AsyncRequest *req = (AsyncRequest *) work->data;
-  req->error = substituteFont(&req->result, req->postscriptName, req->substitutionString);
-}
-
+// -----------------------------------------------------------------------------
+// Substitute Fonts
+// -----------------------------------------------------------------------------
 template<bool async>
-NAN_METHOD(substituteFont) {
-  if (info.Length() < 1 || !info[0]->IsString())
-    return Nan::ThrowTypeError("Expected postscript name");
+Napi::Value FontManagerInterface::substituteFont(const Napi::CallbackInfo& info) {
+  Napi::Env const &env = info.Env();
+  FontManagerImpl *impl = env.GetInstanceData<FontManagerImpl>();
 
-  if (info.Length() < 2 || !info[1]->IsString())
-    return Nan::ThrowTypeError("Expected substitution string");
+  if (info.Length() < 1 || !info[0].IsString()) {
+    return throwError<async>(env, Napi::TypeError::New(env, "Expected postscript name"));
+  }
+  std::string postscriptName = info[0].As<Napi::String>();
```

Comment:
better to add this line after all the checks have been done, to prevent executing code that might not be used if the next line fails/throws

Reply:
I actually did this intentionally for code-cleanliness reasons but you're probably right.